<h3>Imports</h3>

In [2]:
import duckdb
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

<h3>Extracting data from DB</h3>

In [4]:
con = duckdb.connect("../../../dataManagementBackbone/data/exploitation/crimesPrices.db")
df = con.execute("SELECT * FROM crimesPrices").df()

<h3>Scaling numeric attributes</h3>

We are only interesting in scaling the numeric variables that will be used as explanatory varibales in the regression model.

In [5]:
cols = df.iloc[:,7:].select_dtypes(np.number).columns
df[cols] = (df[cols] - df[cols].min()) / (df[cols].max() - df[cols].min())

<h3>Correlation matrix</h3>

In [34]:
corr = df.corr()
corr.style.background_gradient(cmap='coolwarm')

From this correlation matrix, we can see that all crimes have a negative effect on the average price. We should mention that correlation does not mean causation. Therefore, we cannot say that an increase in the number of crimes affects the price. We can say that there seems to be a relationship between the number of crimes and the average price of properties per district.

We can also observe how the number of employed residents, shows a positive relationship with price. A negative relationship can be seen between price and the number of unemplyed. This is to be expected as ususally, the areas with better more expensive houses are populated by people with well-paid jobs.

<h2>District Only<h2>

<h3>Splitting dataset</h3>

In [36]:
y = df.AveragePrice
X = pd.get_dummies(df.iloc[:, 1])


In [37]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

<h3>Mutiple Linear Regression</h3>

In [38]:
reg = LinearRegression().fit(X_train, y_train)
print('R-squared train')
print(reg.score(X_train, y_train))

print('\nR-squared test')
print(reg.score(X_test, y_test))

print('\nPrediction error')
print('\nMean Absolute Error')
print(np.sum(abs(y_test - reg.predict(X_test))) / len(y_test))
print('\nRoot Mean Squared Error')
print(math.sqrt(np.sum((y_test - reg.predict(X_test))*(y_test - reg.predict(X_test))) / len(y_test)))

print('\nCoefficients')
coef = pd.DataFrame([reg.coef_], columns=list(X.columns))
pd.set_option('display.max_columns', None)
display(coef)
print(reg.intercept_)


R-squared train
0.8729482303470458

R-squared test
0.865126736169055

Prediction error

Mean Absolute Error
9311.607142857143

Root Mean Squared Error
11209.980506864662

Coefficients


Boston  East Lindsey       Lincoln  North Kesteven  South Holland  \
0  4.822189e+16  4.822189e+16  4.822189e+16    4.822189e+16   4.822189e+16   

   South Kesteven  West Lindsey  
0    4.822189e+16  4.822189e+16

-4.822188791563829e+16


<h2>Crimes Only<h2>

<h3>Splitting dataset</h3>

In [42]:
y = df.AveragePrice
dropCols = [s for s in df.columns.to_list() if 'Lon' in s or 'Lat' in s] + ['Month', 'District', 'AveragePrice', 'AveragePriceDetached', 'AveragePriceSemiDetached', 'AveragePriceTerraced', 'AveragePriceFlatOrMaisonette', 'Economically inactive', 'Employed', 'Unemployed']
X = df.drop(dropCols, axis=1)
numCols = X.select_dtypes('number').columns.to_list()
X[numCols] = (X[numCols]-X[numCols].min())/(X[numCols].max()-X[numCols].min())
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

<h3>Mutiple Linear Regression</h3>

In [43]:
reg = LinearRegression().fit(X_train, y_train)
print('R-squared train')
print(reg.score(X_train, y_train))

print('\nR-squared test')
print(reg.score(X_test, y_test))

print('\nPrediction error')
print('\nMean Absolute Error')
print(np.sum(abs(y_test - reg.predict(X_test))) / len(y_test))
print('\nRoot Mean Squared Error')
print(math.sqrt(np.sum((y_test - reg.predict(X_test))*(y_test - reg.predict(X_test))) / len(y_test)))

print('\nCoefficients')
coef = pd.DataFrame([reg.coef_], columns=list(X.columns))
pd.set_option('display.max_columns', None)
display(coef)
print(reg.intercept_)


R-squared train
0.7031696313644349

R-squared test
0.6424301534571597

Prediction error

Mean Absolute Error
15410.342013319987

Root Mean Squared Error
18252.500450726588

Coefficients


Anti-social behaviour Count  Criminal damage and arson Count  \
0                  5864.723172                    -34144.692765   

   Violence and sexual offences Count  Public order Count  Other crime Count  \
0                       -31116.954056       -27531.345193      -79072.294104   

     Theft Count  Burglary Count  Vehicle crime Count   Drugs Count  \
0  152846.969036   150103.881478          3381.403127 -74359.511286   

   Possession of weapons Count  Shoplifting Count  Robbery Count  
0                 24557.327927      -57783.959671   -2886.595887

210809.38282499017


<h2>Crimes And District<h2>

<h3>Splitting dataset</h3>

In [44]:
y = df.AveragePrice
dropCols = [s for s in df.columns.to_list() if 'Lon' in s or 'Lat' in s] + ['Month', 'AveragePrice', 'AveragePriceDetached', 'AveragePriceSemiDetached', 'AveragePriceTerraced', 'AveragePriceFlatOrMaisonette', 'Economically inactive', 'Employed', 'Unemployed']
X = df.drop(dropCols, axis=1)
numCols = X.select_dtypes('number').columns.to_list()
X[numCols] = (X[numCols]-X[numCols].min())/(X[numCols].max()-X[numCols].min())
X = X.join(pd.get_dummies(X['District'])).iloc[:, 1:]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

<h3>Mutiple Linear Regression</h3>

In [45]:
reg = LinearRegression().fit(X_train, y_train)
print('R-squared train')
print(reg.score(X_train, y_train))

print('\nR-squared test')
print(reg.score(X_test, y_test))

print('\nPrediction error')
print('\nMean Absolute Error')
print(np.sum(abs(y_test - reg.predict(X_test))) / len(y_test))
print('\nRoot Mean Squared Error')
print(math.sqrt(np.sum((y_test - reg.predict(X_test))*(y_test - reg.predict(X_test))) / len(y_test)))

print('\nCoefficients')
coef = pd.DataFrame([reg.coef_], columns=list(X.columns))
pd.set_option('display.max_columns', None)
display(coef)
print(reg.intercept_)


R-squared train
0.9713456860409166

R-squared test
0.9720420595933867

Prediction error

Mean Absolute Error
3854.4150345288353

Root Mean Squared Error
5103.8101493218155

Coefficients


Anti-social behaviour Count  Criminal damage and arson Count  \
0                -54170.978457                    -54018.511559   

   Violence and sexual offences Count  Public order Count  Other crime Count  \
0                        32859.713305        29635.236108       -2540.452516   

    Theft Count  Burglary Count  Vehicle crime Count   Drugs Count  \
0  34965.214455    39043.973354          6315.495307  49848.177785   

   Possession of weapons Count  Shoplifting Count  Robbery Count  \
0                 17095.954928       24790.519176   19329.219176   

         Boston  East Lindsey       Lincoln  North Kesteven  South Holland  \
0 -41970.214869    338.843594 -70960.889413    52079.735511   18471.016053   

   South Kesteven  West Lindsey  
0    34132.946547   7908.562576

176209.1137438481


<h2>Crimes, District and Month<h2>

<h3>Splitting dataset</h3>

In [8]:
X.columns

Index(['Anti-social behaviour Count', 'Criminal damage and arson Count',
       'Violence and sexual offences Count', 'Public order Count',
       'Other crime Count', 'Theft Count', 'Burglary Count',
       'Vehicle crime Count', 'Drugs Count', 'Possession of weapons Count',
       'Shoplifting Count', 'Robbery Count', '2021-06', '2021-07', '2021-08',
       '2021-09', '2021-10', '2021-11', '2021-12', '2022-01', '2022-02',
       '2022-03', '2022-04', '2022-05', 'Boston', 'East Lindsey', 'Lincoln',
       'North Kesteven', 'South Holland', 'South Kesteven', 'West Lindsey'],
      dtype='object')

In [6]:
y = df.AveragePrice
dropCols = [s for s in df.columns.to_list() if 'Lon' in s or 'Lat' in s] + ['AveragePrice', 'AveragePriceDetached', 'AveragePriceSemiDetached', 'AveragePriceTerraced', 'AveragePriceFlatOrMaisonette', 'Economically inactive', 'Employed', 'Unemployed']
X = df.drop(dropCols, axis=1)
numCols = X.select_dtypes('number').columns.to_list()
X[numCols] = (X[numCols]-X[numCols].min())/(X[numCols].max()-X[numCols].min())
X = X.join(pd.get_dummies(X['Month'])).iloc[:, 1:]
X = X.join(pd.get_dummies(X['District'])).iloc[:, 1:]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

<h3>Mutiple Linear Regression</h3>

In [7]:
reg = LinearRegression().fit(X_train, y_train)
print('R-squared train')
print(reg.score(X_train, y_train))

print('\nR-squared test')
print(reg.score(X_test, y_test))

print('\nPrediction error')
print('\nMean Absolute Error')
print(np.sum(abs(y_test - reg.predict(X_test))) / len(y_test))
print('\nRoot Mean Squared Error')
print(math.sqrt(np.sum((y_test - reg.predict(X_test))*(y_test - reg.predict(X_test))) / len(y_test)))

print('\nCoefficients')
coef = pd.DataFrame([reg.coef_], columns=list(X.columns))
pd.set_option('display.max_columns', None)
display(coef)
print(reg.intercept_)

R-squared train
0.9810522885849212

R-squared test
0.9782372748691788

Prediction error

Mean Absolute Error
3285.891351602863

Root Mean Squared Error
4502.964905076852

Coefficients


Anti-social behaviour Count  Criminal damage and arson Count  \
0                -15642.150937                    -59420.591078   

   Violence and sexual offences Count  Public order Count  Other crime Count  \
0                        37511.224904         34485.22622      -12564.553601   

    Theft Count  Burglary Count  Vehicle crime Count  Drugs Count  \
0  33936.649957    37157.035088          8859.543693  31421.53471   

   Possession of weapons Count  Shoplifting Count  Robbery Count      2021-06  \
0                 19462.396075       -8186.182737    21422.09707 -7428.832174   

       2021-07      2021-08      2021-09      2021-10      2021-11  \
0 -7470.609321 -9634.445416 -5524.186314  4510.286653 -2564.808878   

       2021-12      2022-01      2022-02      2022-03      2022-04  \
0  1570.896158 -1143.004155  3527.194264  5042.613487  7726.627063   

        2022-05        Boston  East Lindsey       Lincoln  North Kesteven  \
0  11388.268633 -34430.038135  -5593.976478 -58824.369817     42246.01019   

   South Holland  South Kesteven  West Lindsey  
0   11453.618055    36432.764124   8715.992061

180170.4873403276


<h2>Crimes, District and Economic Activity<h2>

<h3>Splitting dataset</h3>

In [8]:
y = df.AveragePrice
dropCols = [s for s in df.columns.to_list() if 'Lon' in s or 'Lat' in s] + ['Month', 'AveragePrice', 'AveragePriceDetached', 'AveragePriceSemiDetached', 'AveragePriceTerraced', 'AveragePriceFlatOrMaisonette']
X = df.drop(dropCols, axis=1)
numCols = X.select_dtypes('number').columns.to_list()
X[numCols] = (X[numCols]-X[numCols].min())/(X[numCols].max()-X[numCols].min())
X = X.join(pd.get_dummies(X['District'])).iloc[:, 1:]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

<h3>Mutiple Linear Regression</h3>

In [9]:
reg = LinearRegression().fit(X_train, y_train)
print('R-squared train')
print(reg.score(X_train, y_train))

print('\nR-squared test')
print(reg.score(X_test, y_test))

print('\nPrediction error')
print('\nMean Absolute Error')
print(np.sum(abs(y_test - reg.predict(X_test))) / len(y_test))
print('\nRoot Mean Squared Error')
print(math.sqrt(np.sum((y_test - reg.predict(X_test))*(y_test - reg.predict(X_test))) / len(y_test)))

print('\nCoefficients')
coef = pd.DataFrame([reg.coef_], columns=list(X.columns))
pd.set_option('display.max_columns', None)
display(coef)
print(reg.intercept_)

R-squared train
0.9724487886794692

R-squared test
0.9707748970123579

Prediction error

Mean Absolute Error
3971.7326370234478

Root Mean Squared Error
5218.190729651223

Coefficients


Economically inactive     Employed   Unemployed  \
0           35609.960504  7970.060179  4840.013595   

   Anti-social behaviour Count  Criminal damage and arson Count  \
0                 -54753.15264                    -59252.076544   

   Violence and sexual offences Count  Public order Count  Other crime Count  \
0                        41516.351022         39302.04255       -3954.208715   

    Theft Count  Burglary Count  Vehicle crime Count  Drugs Count  \
0  36902.546052    37758.292531         10609.488627  53906.87695   

   Possession of weapons Count  Shoplifting Count  Robbery Count  \
0                 17895.919446       24325.940077   17305.051679   

         Boston  East Lindsey       Lincoln  North Kesteven  South Holland  \
0 -22330.533785  -23111.36193 -73386.194426    52565.190434   26796.934968   

   South Kesteven  West Lindsey  
0    22713.162443  16752.802295

151436.38054993085


<h2>Crimes, District, Month and Economic Activity<h2>

<h3>Splitting dataset</h3>

In [62]:
y = df.AveragePrice
dropCols = [s for s in df.columns.to_list() if 'Lon' in s or 'Lat' in s] + ['AveragePrice', 'AveragePriceDetached', 'AveragePriceSemiDetached', 'AveragePriceTerraced', 'AveragePriceFlatOrMaisonette']
X = df.drop(dropCols, axis=1)
numCols = X.select_dtypes('number').columns.to_list()
X[numCols] = (X[numCols]-X[numCols].min())/(X[numCols].max()-X[numCols].min())
X = X.join(pd.get_dummies(X['Month'])).iloc[:, 1:]
X = X.join(pd.get_dummies(X['District'])).iloc[:, 1:]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

<h3>Mutiple Linear Regression</h3>

In [63]:
reg = LinearRegression().fit(X_train, y_train)
print('R-squared train')
print(reg.score(X_train, y_train))

print('\nR-squared test')
print(reg.score(X_test, y_test))

print('\nPrediction error')
print('\nMean Absolute Error')
print(np.sum(abs(y_test - reg.predict(X_test))) / len(y_test))
print('\nRoot Mean Squared Error')
print(math.sqrt(np.sum((y_test - reg.predict(X_test))*(y_test - reg.predict(X_test))) / len(y_test)))

print('\nCoefficients')
coef = pd.DataFrame([reg.coef_], columns=list(X.columns))
pd.set_option('display.max_columns', None)
display(coef)
print(reg.intercept_)

R-squared train
0.9824114535373675

R-squared test
0.9769292290231852

Prediction error

Mean Absolute Error
3610.8577424125288

Root Mean Squared Error
4636.315455719247

Coefficients


Economically inactive      Employed    Unemployed  \
0           28605.647184  15064.147998 -27235.784651   

   Anti-social behaviour Count  Criminal damage and arson Count  \
0                -14398.219366                    -73191.798097   

   Violence and sexual offences Count  Public order Count  Other crime Count  \
0                        58794.409149        56333.162219      -20644.971461   

    Theft Count  Burglary Count  Vehicle crime Count  Drugs Count  \
0  34164.139692    34255.223278         15653.316575  34005.63893   

   Possession of weapons Count  Shoplifting Count  Robbery Count      2021-06  \
0                  18839.20902        -14394.5125    18200.59283 -8434.038122   

       2021-07      2021-08      2021-09      2021-10      2021-11    2021-12  \
0 -7460.945414 -9918.819776 -3877.278219  4946.516612 -2468.996053  849.78894   

      2022-01      2022-02      2022-03      2022-04       2022-05  \
0 -2324.59527  3775.390677  5003.834473  8344.633868  11564.508283   

         Boston  East Lindsey       Lincoln  North Kesteven  South Holland  \
0 -27399.583804 -14177.999739 -39569.417306    30835.832021    7859.524834   

   South Kesteven  West Lindsey  
0    29416.318464   13035.32553

170325.98578598586


Despite the error being larger, we will keep the model with the months because it does have a very small difference in correlation with the average price over time.

In [11]:
con.close()

<h2>Conclusion</h2>
Best model: Crimes, district and month

Disclaimer: The metrics will vary with every new execution of the whole data science project. However, the general behaviour is still the same.

We begin with creating a model using just the districts as the predictors. This model performed relatively well (0.87 R-squared) for the small number of predictors. This shows that the main affect on the average price comes from the district the property is in. This makes sense because there areas that are going to be more expensive due to the properties being bigger.

The second model, uses the counts of all the crimes as predictors. This model did not perform as well as the district model. However, it does show that there exists a relationship between the number of crimes and the average price.

Next, we create a model using the districts and the crime counts. We did not use the longitude or latitude of the crimes because they did not prove useful. The resulting model is the best performing yet with an R-squared of 0.97 for both training and testing, meaning that there is no ovrefitting or underfitting. This is very promising. With a RMSE of 5103, we can say that this model is quite accurate. From now on, we will continue adding different variables to perfect the model without overfitting it.

We created models with:
<ul>
    <li>Crimes, District, Month</li>
    <ul>
        <li>R-squared training: 0.9819</li>
        <li>R-squared testing: 0.9782</li>
        <li>RMSE: 4502.96</li>
    </ul>
    <li>Crimes, District, Economic Status</li>
    <ul>
        <li>R-squared training: 0.9724</li>
        <li>R-squared testing: 0.9707</li>
        <li>RMSE: 5218.19</li>
    </ul>
    <li>Crimes, District, Month, Economic Status</li>
    <ul>
        <li>R-squared training: 0.9824</li>
        <li>R-squared testing: 0.9769</li>
        <li>RMSE: 4636.32</li>
    </ul>
</ul>

All of the models above, have a good performance. However, the model with the least RMSE and the best balance between the multiple correlation coefficient, R-squared, for both training and testing is the model with crimes, district and month. Therefore, we will select this model as our final version. This means that the data discovery task wasn't useful for our analysis question. However, it could prove useful for other analysis and we could showcase that our data management backbone can scale easily with new data sources. We should mention that the data sources must fulfill certain requirements, such as, sharing the same months and districts as our original data. If this isn't fulfilled we would not be able to perform the join operation.



Finally, we should mention that the data set is relatively small and it uses aggreagated values which don't change much month to month. This makes quite easy to predict a correct value for each district. Therefore, this model can only really be used to predict the average price per district over the months. It cannot be used to predict the price of a particular property. For future improvements, we would need to change the main dataset of prices. We would need to find/create a data set with prices of individual properties with attributes such as the district, size, number of rooms, outdoor area... The resulting model from this dataset would have the potential to be more useful than the one we have developed if it had a good performance.